gapir/cc/memory_allocator.cpp

```diff
@@ -0,0 +1,506 @@
+/*
+ * Copyright (C) 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <cstring>
+#include <iostream>
+#include <set>
+#include <sstream>
+
+#include <sys/time.h>
+
+#include "core/cc/log.h"
+#include "memory_allocator.h"
+
+namespace gapir {
+
+std::map<unsigned int,
+         std::map<unsigned char*, MemoryAllocator::MemoryRegion>::iterator>
+    MemoryAllocator::Handle::_dummyMap;
+
+std::unique_ptr<MemoryAllocator> MemoryAllocator::create(size_t heapSize) {
+  return std::unique_ptr<MemoryAllocator>(new MemoryAllocator(heapSize));
+}
+
+MemoryAllocator::MemoryAllocator(size_t heapSize)
+    : heapSize_(heapSize),
+      heap_(new unsigned char[heapSize]),
```

Comment:
This throws `std::bad_alloc` if the memory allocation fails, you probably want this instead:
`heap_(new (std::nothrow) unsigned char[heapSize]),`
to get a null pointer if the mem alloc fails?

see http://www.cplusplus.com/reference/new/operator%20new/

Reply:
This throws an exception, correct. The memory allocator is created inside a while() try {} catch{} in main.cpp.